### ISSUE TITLE
ValueError when client ssh config references unimplemented algorithms

### ISSUE THREAD

**Issue by matthazinski** [ronf/asyncssh#359]
When one algorithm (e.g. MAC, KexAlgorithm) in a user's client config is not supported by asyncssh, `_expand_algs()` will raise a ValueError, even if there are other supported algorithms in a user's config which are compatible with both asyncssh and the remote sshd.

This is the same behavior as [OpenSSH](https://github.com/openssh/openssh-portable/blob/31d8d231eb9377df474746a822d380c5d68d7ad6/servconf.c#L1809), but I'm wondering whether this is desirable. Currently, it is not possible for me to reuse a client config between the `master` branch of asyncssh and openssh-8.5 when either:

  * I'd like to use bleeding-edge crypto which it not yet implemented in asyncssh, e.g. the `sntrup761x25519-sha512@openssh.com` `KexAlgorithm`); or
  * My system does not have libnettle available for asyncssh, but OpenSSH has umac algorithm support compiled.

As an example repro:
```
[mhazinsk@host asyncssh]$ cat example.py
import asyncio, asyncssh
import tempfile, traceback

async def run_client(config):
    async with asyncssh.connect('myhost.example.com', config=[config]) as conn:
        result = await conn.run('id', check=True)
        print(result.stdout, end='')

kexblock = b'KexAlgorithms sntrup761x25519-sha512@openssh.com,curve25519-sha256,diffie-hellman-group14-sha1,diffie-hellman-group-exchange-sha256\n'

for config_body in [kexblock]:
    with tempfile.NamedTemporaryFile(delete=False) as f:
        f.write(config_body)
        f.close()
        try:
            asyncio.get_event_loop().run_until_complete(run_client(config=f.name))
        except Exception:
            print(traceback.format_exc())
```

```
[mhazinsk@host asyncssh]$ python3 example.py
Traceback (most recent call last):
  File "/home/mhazinsk/local/asyncssh/example.py", line 23, in <module>
    asyncio.get_event_loop().run_until_complete(run_client(config=f.name))
  File "/usr/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/home/mhazinsk/local/asyncssh/example.py", line 6, in run_client
    async with asyncssh.connect('myhost.example.com', config=[config]) as conn:
  File "/home/mhazinsk/local/asyncssh/asyncssh/misc.py", line 220, in __aenter__
    self._result = await self._coro
  File "/home/mhazinsk/local/asyncssh/asyncssh/connection.py", line 6466, in connect
    options = SSHClientConnectionOptions(options, config=config, host=host,
  File "/home/mhazinsk/local/asyncssh/asyncssh/connection.py", line 5395, in __init__
    super().__init__(options=options, last_config=last_config, **kwargs)
  File "/home/mhazinsk/local/asyncssh/asyncssh/misc.py", line 268, in __init__
    self.prepare(**self.kwargs)
  File "/home/mhazinsk/local/asyncssh/asyncssh/connection.py", line 5855, in prepare
    super().prepare(config, client_factory or SSHClient, client_version,
  File "/home/mhazinsk/local/asyncssh/asyncssh/connection.py", line 5425, in prepare
    _validate_algs(config, kex_algs, encryption_algs, mac_algs,
  File "/home/mhazinsk/local/asyncssh/asyncssh/connection.py", line 380, in _validate_algs
    kex_algs = _select_algs('key exchange', kex_algs, get_kex_algs(),
  File "/home/mhazinsk/local/asyncssh/asyncssh/connection.py", line 341, in _select_algs
    algs = _expand_algs(alg_type, algs, possible_algs, default_algs)
  File "/home/mhazinsk/local/asyncssh/asyncssh/connection.py", line 315, in _expand_algs
    raise ValueError('"%s" matches no valid %s algorithms' %
ValueError: "sntrup761x25519-sha512@openssh.com" matches no valid key exchange algorithms
```

**Comment by ronf** [comment]
Thanks for the report.

I'm a bit torn on this one. While I could make AsyncSSH more forgiving here, that would only help in cases where OpenSSH implemented something that AsyncSSH didn't. There are a few such examples (like the one you listed here), but there are a much larger number of algorithms supported by AsyncSSH but not OpenSSH, and you'd still run into a problem trying to share a config file in that case even if I did relax the AsyncSSH matching.

I also worry a bit here that just silently ignoring a non-matching algorithm might prevent us from catching a typo in an algorithm name that would be caught today.

If I were to do this, I think I would definitely want to make it apply ONLY to algorithms read from the config file. That way, any algorithm names passed directly via AsyncSSH arguments would get the full checking applied on them. The good news is that it isn't difficult to make that conditional. It would look something like:

```patch
--- a/asyncssh/connection.py
+++ b/asyncssh/connection.py
@@ -300,7 +300,7 @@ def _validate_version(version):
     return version


-def _expand_algs(alg_type, algs, possible_algs, default_algs):
+def _expand_algs(alg_type, algs, possible_algs, default_algs, strict_match):
     """Expand the set of allowed algorithms"""

     if algs[:1] in '^+-':
@@ -317,7 +317,7 @@ def _expand_algs(alg_type, algs, possible_algs, default_algs):
         matches = [alg for alg in possible_algs
                    if pattern.matches(alg.decode('ascii'))]

-        if not matches:
+        if not matches and strict_match:
             raise ValueError('"%s" matches no valid %s algorithms' %
                              (pat, alg_type))

@@ -339,12 +339,16 @@ def _select_algs(alg_type, algs, possible_algs, default_algs,

     if algs == ():
         algs = config_algs
+        strict_match = False
+    else:
+        strict_match = True

     if algs == ():
         return default_algs
     elif algs:
         if isinstance(algs, str):
-            algs = _expand_algs(alg_type, algs, possible_algs, default_algs)
+            algs = _expand_algs(alg_type, algs, possible_algs,
+                                default_algs, strict_match)
         else:
             algs = [alg.encode('ascii') for alg in algs]

```

If this were available, how would you handle algorithms available in AsyncSSH but not in OpenSSH?

**Comment by ronf** [comment]
I never heard back from you on my last question above, but I've gone ahead and checked in this change (along with updated unit tests) in commit 6283180 in the "develop" branch, and will be including it in the next release.

**Comment by ronf** [comment]
This change is now available in AsyncSSH 2.6.0.
